binderhub/build_local.py

```diff
@@ -38,28 +70,55 @@ def _execute_cmd(cmd, capture=False, **kwargs):
     # This should behave the same as .readline(), but splits on `\r` OR `\n`,
     # not just `\n`.
     buf = []
+    q = queue.Queue()
+
+    def read_to_queue(proc, capture, q):
+        try:
+            for c in iter(partial(proc.stdout.read, 1), b""):
```

Comment:
Question: why not use asyncio subprocesses with a future to signal the break?

```python
break_future = input_arg
p = asyncio.create_subprocess_exec(cmd)
read_future = p.stdout.readline()

while not break_future.done():
    await asyncio.wait([read_future, break_future], return_when=asyncio.FIRST_COMPLETED)
    if read_future.done():
        # the read completed
        line = await read_future
        yield line
        # schedule next read
        if line and not cancel_future.done():
            read_future = p.stdout.readline()
```

This would make it an async generator (`async for line in _execute_cmd(...)`).

Running a threadsafe put for every _character_ is going to get pretty inefficient when repo2docker produces a lot of output. This is less important if localr2d is only used for testing, but if we want to move to prod, there are probably more efficient ways.

Reply:
I see this is avoiding a change to make Build.submit itself a coroutine, which we should perhaps do in the future.

In that case, I think we can switch the `read_to_queue` to call `proc.stdout.readline()` since it doesn't matter anymore if it blocks. It will always unblock when EOF arrives after the process has been terminated.

Reply:
Thanks for reviewing! I've changed it to iterate through `proc.stdout`- which I think is equivalent to your suggestion, and it's a lot cleaner now. The original character-by-character handling was copied from repo2docker:
https://github.com/jupyterhub/repo2docker/blob/4da2e1e71565b26a4bf1d0f0d26ae7c2373a1fd7/repo2docker/utils.py#L34-L46

@yuvipanda and I have discussed defining the interface for the build/launch classes to make the K8s parameters optional, making this function async could be part of that. If you think it's better I'm happy to leave this PR until that refactoring is done.

Reply:
👍 to doing that as a separate task